scalding-core/src/main/scala/com/twitter/scalding/typed/memory_backend/MemoryBackend.scala

```diff
@@ -0,0 +1,527 @@
+  def setter[U <: T] = ???
+  def writeFrom(pipe: Pipe)(implicit flowDef: FlowDef, mode: Mode): Pipe = ???
+}
+
+class MemoryWriter(mem: MemoryMode) extends Writer {
+
+  import MemoryPlanner.{ Memo, Op }
+  import TypedPipe._
+
+  def start(): Unit = ()
+  /**
+   * This is called by an Execution to end processing
+   */
+  def finished(): Unit = ()
+
+  def plan[T](m: Memo, tp: TypedPipe[T]): (Memo, Op[T]) =
+    m.plan(tp) {
+      tp match {
+        case cp@CrossPipe(_, _) =>
+          plan(m, cp.viaHashJoin)
+
+        case CrossValue(left, EmptyValue) => (m, Op.empty)
+        case CrossValue(left, LiteralValue(v)) =>
+          val (m1, op) = plan(m, left)
+          (m1, op.flatMap { a => Iterator.single((a, v)) })
+        case CrossValue(left, ComputedValue(right)) =>
+          plan(m, CrossPipe(left, right))
+        case DebugPipe(p) =>
+          // There is really little that can be done here but println
+          plan(m, p.map { t => println(t); t })
+
+        case EmptyTypedPipe =>
+          // just use an empty iterable pipe.
+          // Note, rest is irrelevant
+          (m, Op.empty[T])
+
+        case fk@FilterKeys(_, _) =>
+          def go[K, V](node: FilterKeys[K, V]): (Memo, Op[(K, V)]) = {
+            val FilterKeys(pipe, fn) = node
+            val (m1, op) = plan(m, pipe)
+            (m1, op.flatMap { case (k, v) => if (fn(k)) { (k, v) :: Nil } else Nil })
+          }
+          go(fk)
+
+        case f@Filter(_, _) =>
+          def go[T](f: Filter[T]): (Memo, Op[T]) = {
+            val Filter(p, fn) = f
+            val (m1, op) = plan(m, f)
+            (m1, op.flatMap { t => if (fn(t)) t :: Nil else Nil })
+          }
+          go(f)
+
+        case f@FlatMapValues(_, _) =>
+          def go[K, V, U](node: FlatMapValues[K, V, U]) = {
+            // don't capture node, which is a TypedPipe, which we avoid serializing
+            val fn = node.fn
+            val (m1, op) = plan(m, node.input)
+            (m1, op.flatMap { case (k, v) => fn(v).map((k, _)) })
+          }
+
+          go(f)
+
+        case FlatMapped(prev, fn) =>
+          val (m1, op) = plan(m, prev)
+          (m1, op.flatMap(fn))
+
+        case ForceToDisk(pipe) =>
+          val (m1, op) = plan(m, pipe)
+          (m1, Op.Materialize(op))
+
+        case Fork(pipe) =>
+          val (m1, op) = plan(m, pipe)
+          (m1, Op.Materialize(op))
+
+        case IterablePipe(iterable) =>
+          (m, Op.source(iterable))
+
+        case f@MapValues(_, _) =>
+          def go[K, V, U](node: MapValues[K, V, U]) = {
+            // don't capture node, which is a TypedPipe, which we avoid serializing
+            val mvfn = node.fn
+            val (m1, op) = plan(m, node.input)
+            (m1, op.flatMap { case (k, v) => Iterator.single((k, mvfn(v))) })
+          }
+
+          go(f)
+
+        case Mapped(input, fn) =>
+          val (m1, op) = plan(m, input)
+          (m1, op.flatMap { t => fn(t) :: Nil })
```

Comment:
it might be worth adding a `map` method to avoid creating one list for each element.